### PR TITLE
feat(app): add runtime parameters to ChooseProtocolSlideout

### PR DESCRIPTION
**Pull request by ncdiehl11**
closes [AUTH-246](https://opentrons.atlassian.net/browse/AUTH-246)

# Overview

Add RTP functionality to `ChooseProtocolSlideout` 

# Test Plan

- upload an RTP-containing protocol to the app
- navigate to `DeviceDetails` page 
- select 'Run a protocol' from overflow menu
- select an RTP protocol from the populated slideout
- observe flow exposing RTP parameter input and ability to go back and change protocol
- note that non-RTP-containing protocols show only one step to 'Proceed to setup'

https://github.com/Opentrons/opentrons/assets/47604184/84cdf53d-71c2-4baf-b501-0e36faea3fae

# Changelog

- add RTPs to `ChooseProtocolSlideout`
- send RTP overrides to `useCreateRunFromProtocol` hook

# Review requests

authorship devs

# Risk assessment
low

[AUTH-246]: https://opentrons.atlassian.net/browse/AUTH-246?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/src/organisms/ChooseProtocolSlideout/index.tsx [hunk 1]
```diff
@@ -39,6 +51,7 @@ import { getAnalysisStatus } from '../ProtocolsLanding/utils'
 import type { Robot } from '../../redux/discovery/types'
 import type { StoredProtocolData } from '../../redux/protocol-storage'
 import type { State } from '../../redux/types'
+import type { RunTimeParameter } from '@opentrons/shared-data'
```
Comment (koji) [inline app/src/organisms/ChooseProtocolSlideout/index.tsx]: this line can be before `import type { Robot } from '../../redux/discovery/types'`

File: app/src/organisms/ChooseProtocolSlideout/index.tsx [hunk 2]
```diff
@@ -72,6 +85,30 @@ export function ChooseProtocolSlideoutComponent(
...
+  React.useEffect(() => {
+    setRunTimeParametersOverrides(
+      selectedProtocol?.mostRecentAnalysis?.runTimeParameters ?? []
+    )
+  }, [selectedProtocol])
+  const runTimeParametersFromAnalysis =
+    selectedProtocol?.mostRecentAnalysis?.runTimeParameters ?? []
+
+  const [currentPage, setCurrentPage] = React.useState<number>(1)
+  const enableRunTimeParametersFF = useFeatureFlag('enableRunTimeParameters')
```
Comment (koji) [inline app/src/organisms/ChooseProtocolSlideout/index.tsx]: Can we move these two lines right after `React.useState<RunTimeParameter[]>([])`?

File: app/src/organisms/ChooseProtocolSlideout/index.tsx [hunk 3]
```diff
@@ -72,6 +85,30 @@ export function ChooseProtocolSlideoutComponent(
...
+  const hasRunTimeParameters =
+    enableRunTimeParametersFF && runTimeParametersFromAnalysis.length > 0
+
+  const [targetProps, tooltipProps] = useHoverTooltip()
+
+  console.log(runTimeParametersOverrides)
```
Comment (koji) [inline app/src/organisms/ChooseProtocolSlideout/index.tsx]: need to keep for debugging purposes?

File: app/src/organisms/ChooseProtocolSlideout/index.tsx [hunk 4]
```diff
@@ -72,6 +85,30 @@ export function ChooseProtocolSlideoutComponent(
...
+  }, [selectedProtocol])
+  const runTimeParametersFromAnalysis =
+    selectedProtocol?.mostRecentAnalysis?.runTimeParameters ?? []
+
+  const [currentPage, setCurrentPage] = React.useState<number>(1)
+  const enableRunTimeParametersFF = useFeatureFlag('enableRunTimeParameters')
+  const hasRunTimeParameters =
+    enableRunTimeParametersFF && runTimeParametersFromAnalysis.length > 0
+
+  const [targetProps, tooltipProps] = useHoverTooltip()
```
Comment (koji) [inline app/src/organisms/ChooseProtocolSlideout/index.tsx]: also, this line can be moved up.

File: app/src/organisms/ChooseProtocolSlideout/index.tsx (hunk 1 above)
Comment (ncdiehl11) [reply app/src/organisms/ChooseProtocolSlideout/index.tsx]: thanks— sorry, the auto import always puts it at the bottom and i forget to move it up 😅

File: app/src/organisms/ChooseProtocolSlideout/index.tsx (hunk 3 above)
Comment (ncdiehl11) [reply app/src/organisms/ChooseProtocolSlideout/index.tsx]: no need, going to remove now.

File: app/src/organisms/ChooseProtocolSlideout/index.tsx [hunk 5]
```diff
@@ -141,10 +181,234 @@ export function ChooseProtocolSlideoutComponent(
...
+                  return {
+                    ...parameter,
+                    value:
+                      dropdownOptions.find(option => option.value === choice)
+                        ?.value ?? parameter.default,
+                  }
+                }
+                return parameter
+              })
+              if (setRunTimeParametersOverrides != null) {
```
Comment (koji) [inline app/src/organisms/ChooseProtocolSlideout/index.tsx]: question

what is the reason to check useState set method?
I think this is always `true`.

File: app/src/organisms/ChooseProtocolSlideout/index.tsx [hunk 6]
```diff
@@ -141,10 +181,234 @@ export function ChooseProtocolSlideoutComponent(
...
+      <Flex justifyContent={JUSTIFY_END}>
+        <LinkComponent
+          textAlign={TYPOGRAPHY.textAlignRight}
+          css={
+            isRestoreDefaultsLinkEnabled ? ENABLED_LINK_CSS : DISABLED_LINK_CSS
+          }
+          onClick={() => {
+            const clone = runTimeParametersOverrides.map(parameter => ({
+              ...parameter,
+              value: parameter.default,
```
Comment (koji) [inline app/src/organisms/ChooseProtocolSlideout/index.tsx:338]: looks fine but I noticed this case.
float case isn't reset as well as others.

https://github.com/Opentrons/opentrons/assets/474225/ad0f4ceb-bdef-45b2-bd17-0c34e863463b

File: app/src/organisms/ChooseProtocolSlideout/index.tsx (hunk 5 above)
Comment (ncdiehl11) [reply app/src/organisms/ChooseProtocolSlideout/index.tsx]: artifact from copypasta. sorry— removing now

File: app/src/organisms/ChooseProtocolSlideout/index.tsx (hunk 6 above)
Comment (ncdiehl11) [reply app/src/organisms/ChooseProtocolSlideout/index.tsx:338]: Interesting, the 6.500...01 is evaluating as equal to 6.5. not sure if this requires a fix? What do you think?

File: app/src/organisms/ChooseProtocolSlideout/index.tsx [hunk 7]
```diff
@@ -141,10 +181,234 @@ export function ChooseProtocolSlideoutComponent(
...
+  )
+
+  const singlePageFooter = (
+    <PrimaryButton
+      onClick={handleProceed}
+      disabled={isCreatingRun || selectedProtocol == null}
+      width="100%"
+    >
+      {isCreatingRun ? (
+        <Icon name="ot-spinner" spin size={SIZE_1} />
```
Comment (koji) [inline app/src/organisms/ChooseProtocolSlideout/index.tsx]: Sorry, I think we haven't shared the story of `SIZE_number`
SIZE consts haven't been updated for a long time and we decided to use the actual number for `size`. 

```suggestion
        <Icon name="ot-spinner" spin size="1rem" />
```

File: app/src/organisms/ChooseProtocolSlideout/index.tsx [hunk 8]
```diff
@@ -401,3 +659,18 @@ function StoredProtocolList(props: StoredProtocolListProps): JSX.Element {
...
+`
+
+const DISABLED_LINK_CSS = css`
+  ${TYPOGRAPHY.linkPSemiBold}
+  color: ${COLORS.grey40};
+  cursor: default;
+
+  &:hover {
+    color: ${COLORS.grey40};
+  }
```
Comment (koji) [inline app/src/organisms/ChooseProtocolSlideout/index.tsx:666]: I think the code itself makes sense, but hover would be prevented by tooltip.
So maybe we can remove these lines?  🤔 
What do you think?

File: app/src/organisms/ChooseProtocolSlideout/index.tsx (hunk 6 above)
Comment (koji) [reply app/src/organisms/ChooseProtocolSlideout/index.tsx:338]: I was interested in that. 
I think this is a super rare case. We don't have to deal with this.

File: app/src/organisms/ChooseProtocolSlideout/index.tsx (hunk 8 above)
Comment (ncdiehl11) [reply app/src/organisms/ChooseProtocolSlideout/index.tsx:666]: Without the `hover` pseudoclass, we get the following:
Screenshot 2024-04-02 at 4 27 02 PM

File: app/src/organisms/ChooseProtocolSlideout/index.tsx [hunk 9]
```diff
@@ -141,10 +181,226 @@ export function ChooseProtocolSlideoutComponent(
...
+        disabled={isCreatingRun || selectedProtocol == null}
+      >
+        {t('shared:continue_to_param')}
+      </PrimaryButton>
+    ) : (
+      <Flex gridGap={SPACING.spacing8} flexDirection={DIRECTION_ROW}>
+        <SecondaryButton onClick={() => setCurrentPage(1)} width="51%">
+          {t('shared:change_protocol')}
+        </SecondaryButton>
+        <PrimaryButton width="49%" onClick={handleProceed}>
```
Comment (koji) [inline app/src/organisms/ChooseProtocolSlideout/index.tsx:387]: maybe we want to have `disabled` here to handle a case that user put invalid values.
what do you think?

https://www.figma.com/file/aS5owa8kwmA0uzqzkP1BLJ?type=design&node-id=4045-51207&mode=dev#756367329

File: app/src/organisms/ChooseProtocolSlideout/index.tsx (hunk 8 above)
Comment (koji) [reply app/src/organisms/ChooseProtocolSlideout/index.tsx:666]: I see, thank you for the explanation. 
interesting

**Review by koji — APPROVED** [review]
LGTM

I left a comment on confirm values button but that can be updated in a following PR since we are waiting for Sue's feedback.

Can we create a ticket to updating test for this?

File: app/src/organisms/ChooseProtocolSlideout/index.tsx
Comment (koji) [inline app/src/organisms/ChooseProtocolSlideout/index.tsx:1]: need to remove `SIZE_1`

File: app/src/organisms/ChooseProtocolSlideout/index.tsx (hunk 9 above)
Comment (koji) [reply app/src/organisms/ChooseProtocolSlideout/index.tsx:387]: Sue updated the Figma's note and we need to make the button disabled.

Comment (ncdiehl11) [reply app/src/organisms/ChooseProtocolSlideout/index.tsx:387]: Thanks Koji, just saw the note in Figma. I will add a disabled state and add a test for it in a new PR

**Pull request merged**